streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -81,29 +87,34 @@ public void register(final StateRestorer restorer) {
 
         try {
             final ConsumerRecords<byte[], byte[]> records = restoreConsumer.poll(pollTime);
-            final Iterator<TopicPartition> iterator = needsRestoring.keySet().iterator();
-            while (iterator.hasNext()) {
-                final TopicPartition partition = iterator.next();
+
+            for (final TopicPartition partition : needsRestoring) {
                 final StateRestorer restorer = stateRestorers.get(partition);
                 final long pos = processNext(records.records(partition), restorer, endOffsets.get(partition));
                 restorer.setRestoredOffset(pos);
                 if (restorer.hasCompleted(pos, endOffsets.get(partition))) {
                     restorer.restoreDone();
                     endOffsets.remove(partition);
-                    iterator.remove();
+                    completedRestorers.add(partition);
                 }
             }
         } catch (final InvalidOffsetException recoverableException) {
             log.warn("Restoring StreamTasks failed. Deleting StreamTasks stores to recreate from scratch.", recoverableException);
             final Set<TopicPartition> partitions = recoverableException.partitions();
             for (final TopicPartition partition : partitions) {
                 final StreamTask task = active.restoringTaskFor(partition);
-                log.info("Reinitializing StreamTask {}", task);
+                log.info("Reinitializing StreamTask {} for changelog {}", task, partition);
+
+                needsInitializing.remove(partition);
```

Comment:
This is the other fix: whenever we initialize, inside `reinitializeStateStoresForPartitions` the store.init will be called which will call storechangelogger.register() again to put it into the `needsInitializing` set.